### PR TITLE
[16.0][IMP] [account_statement_base account_reconcile_oca] make statements listed and editable

### PR DESCRIPTION
**Pull request by lk-eska**
Currently there is no way to list or edit bank statement lines. this PR brings this feature to account_statement_base account_reconcile_oca modules.

### CODE REVIEW

File: account_reconcile_oca/views/account_bank_statement_line.xml [hunk 1]
```diff
@@ -325,11 +326,12 @@
         <field
             name="context"
         >{'default_journal_id': active_id, 'view_ref': 'account_reconcile_oca.bank_statement_line_form_reconcile_view'}</field>
-        <field name="view_mode">tree</field>
+        <field name="view_mode">tree,kanban</field>
```
Comment (etobella) [inline account_reconcile_oca/views/account_bank_statement_line.xml:329]:
```suggestion
        <field name="view_mode">kanban,tree</field>
```

**Review by etobella — COMMENTED** [review]
Just a small comment and we can proceed

File: account_reconcile_oca/views/account_bank_statement_line.xml [hunk 2]
```diff
@@ -306,11 +306,12 @@
         <field
             name="context"
         >{'default_journal_id': active_id, 'search_default_not_reconciled': True, 'view_ref': 'account_reconcile_oca.bank_statement_line_form_reconcile_view'}</field>
-        <field name="view_mode">tree</field>
+        <field name="view_mode">kanban,tree</field>
         <field
             name="view_ids"
             eval="[(5, 0, 0),
-            (0, 0, {'view_mode': 'tree', 'view_id': ref('bank_statement_line_reconcile_view')})]"
+            (0, 0, {'view_mode': 'tree', 'view_id': ref('account_statement_base.account_bank_statement_line_tree')}),
```
Comment (etobella) [inline account_reconcile_oca/views/account_bank_statement_line.xml]: Change the order, first the kanban please :wink:

File: account_reconcile_oca/views/account_bank_statement_line.xml [hunk 3]
```diff
@@ -325,11 +326,12 @@
         <field
             name="context"
         >{'default_journal_id': active_id, 'view_ref': 'account_reconcile_oca.bank_statement_line_form_reconcile_view'}</field>
-        <field name="view_mode">tree</field>
+        <field name="view_mode">tree,kanban</field>
         <field
             name="view_ids"
             eval="[(5, 0, 0),
-            (0, 0, {'view_mode': 'tree', 'view_id': ref('bank_statement_line_reconcile_view')})]"
+            (0, 0, {'view_mode': 'tree', 'view_id': ref('account_statement_base.account_bank_statement_line_tree')}),
```
Comment (etobella) [inline account_reconcile_oca/views/account_bank_statement_line.xml:333]: Change the order, first the kanban please :wink:

**Review by etobella — COMMENTED** [review]
The order is important, as the default loaded view is tree now

File: account_reconcile_oca/views/account_bank_statement_line.xml (hunk 3 above)
Comment (lk-eska) [reply account_reconcile_oca/views/account_bank_statement_line.xml:333]: I just sent a fix.

**Comment by pedrobaeza** [comment]
I haven't checked yet, but statement lines shouldn't be editable once the statement is published (just in case this is favoring other behavior).

**Comment by etobella** [comment]
@pedrobaeza I had the same concern, but I tested it on runboat and an error was raised when editing reconciled moves :smiley:

**Comment by pedrobaeza** [comment]
Then isn't better to not allow to edit such reconciled entries?

**Comment by lk-eska** [comment]
Odoo doesn't allow you to change reconciled entries so there is no such concern.

The real problem is not having edit/delete capability on the statements. How to change or delete them if you imported wrong? How to set foreign currency and foreign currency amount? Is it possible to do it somewhere else?

**Comment by pedrobaeza** [comment]
> Odoo doesn't allow you to change reconciled entries so there is no such concern.

What I mean is UX. According to what I'm understanding, the interface will let you change data in reconciled lines, but at the saving time, it will rise an error. What I'm talking about is to directly avoid to change such data for the reconciled entries.

**Comment by lk-eska** [comment]
How do we allow editing some of the lines conditionally? (on tree editable mode)  We can improve UX if this is possible. It works on enterprise similarly by the way.

**Comment by etobella** [comment]
In 16 you should do: attrs="{'readonly': [('is_reconciled', '=', True)]}" on the view for each field

**Comment by lk-eska** [comment]
> In 16 you should do: attrs="{'readonly': [('is_reconciled', '=', True)]}" on the view for each field

I added this to this PR.

**Comment by etobella** [comment]
Can you squash your commits and a use a proper commit message?

Thanks :wink:

**Comment by lk-eska** [comment]
There should be "Squash & Merge" option for accepting a PR?

**Comment by pedrobaeza** [comment]
No, the bot can't squash, so it should be done manually.

**Comment by etobella** [comment]
I checked and with this new tree view, we can unlink the items. It is right that we can unlink lines that are not reconciled, but with this we were able to unlink reconciled lines. Can you add a validation on the unlink in order avoid this?

**Comment by HaraldPanten** [comment]
> I checked and with this new tree view, we can unlink the items. It is right that we can unlink lines that are not reconciled, but with this we were able to unlink reconciled lines. Can you add a validation on the unlink in order avoid this?

Hi @lk-eska @pedrobaeza I think that it should be fixed before merging this PR. Deleting reconciled items in the tree view should be forbidden. If you want, we can manage this PR keeping you as author.

Let us know, please.

**Comment by lk-eska** [comment]
> No, the bot can't squash, so it should be done manually.

I squashed them into one commit.

**Comment by lk-eska** [comment]
> > I checked and with this new tree view, we can unlink the items. It is right that we can unlink lines that are not reconciled, but with this we were able to unlink reconciled lines. Can you add a validation on the unlink in order avoid this?
> 
> Hi @lk-eska @pedrobaeza I think that it should be fixed before merging this PR. Deleting reconciled items in the tree view should be forbidden. If you want, we can manage this PR keeping you as author.
> 
> Let us know, please.

Deleting a statement line (unless it is reconciled with an invoice) is standard behavior in enterprise. This is not a bug but a feature and all enterprise users live with it without complaining. I think we are getting beyond the scope of this pull request. Not that I care about authorship but if you think this is required you can create your own PR fixing it once this gets merged.

**Comment by etobella** [comment]
@lk-eska It is already nice to delete not reconciled lines, but right now,we are able to delete reconciled lines.

**Comment by lk-eska** [comment]
I can't do it in runboat if it is reconciled with an invoice:

image

It can be deleted if it is reconciled manually and it makes sense.

**Comment by etobella** [comment]
But only with reconciled items with invoices

You can see here what happens... Obviously, this is not right

https://github.com/OCA/account-reconcile/assets/28590170/845c836b-ee8a-4990-bbd9-255207f672d6

**Comment by lk-eska** [comment]
> But only with reconciled items with invoices
> 
> You can see here what happens... Obviously, this is not right
> movie.mp4

It is deleting the generated account move so I see no problem here. Why bother unreconciling it when you can skip that step?

**Comment by etobella** [comment]
I understand that a reconciled move should never be deleted, as it is posted.

So, IMO it is not a desired behavior.

**Comment by HaraldPanten** [comment]
What about payments related to reconciled account moves like manual ones? For example like the ones that we create in the initial balance, or the ones coming from payrolls? Is that a desired behaviour?

**Comment by lk-eska** [comment]
> What about payments related to reconciled account moves like manual ones? For example like the ones that we create in the initial balance, or the ones coming from payrolls? Is that a desired behaviour?

I haven't tested all possible use cases but I assume it allows deletion if it is not reconciled with another move. It should cover most of your questions. IMO, deleting is not a big deal if it is a standalone entry.

**Comment by HaraldPanten** [comment]
> > What about payments related to reconciled account moves like manual ones? For example like the ones that we create in the initial balance, or the ones coming from payrolls? Is that a desired behaviour?
> 
> I haven't tested all possible use cases but I assume it allows deletion if it is not reconciled with another move. It should cover most of your questions. IMO, deleting is not a big deal if it is a standalone entry.

Hi, I'm starting to think that you're right. I've tested V15 and the functionality is the same as here (I think). In V15 you can delete reconciled bank statement lines as well.

**Comment by pedrobaeza** [comment]
> In V15 you can delete reconciled bank statement lines as well.

AFAIK, that's not possible. What are the steps you do for that?

**Comment by lk-eska** [comment]
I think there is a confusion when we say reconciled. Let me clarify:
1) reconciled with another entry such as invoice (deletion is not possible)
2) directly creating a new entry such as manual expense entry (deletion is possible)

**Comment by HaraldPanten** [comment]
These are my steps in V15:

1. Create a bank statement with a bank statement line of -15$.
2. Validate the statement.
3. Go to the reconciliation view.
4. Make a manual operation an send these -15$ to any expenses account.
5. Go to the operations view (see screenshot).

Captura de Pantalla 2024-03-15 a las 11 53 48

6. Select the reconciled line in the tree view and try to delete it from the tree view.

Captura de Pantalla 2024-03-15 a las 11 54 00

8. Users can delete it without any restriction.

@pedrobaeza @etobella I think it's not correct, but it seems it has been working like that for several years. WDYT?

**Comment by HaraldPanten** [comment]
> I think there is a confusion when we say reconciled. Let me clarify:
> 
> 1. reconciled with another entry such as invoice (deletion is not possible)
> 2. directly creating a new entry such as manual expense entry (deletion is possible)

Yep. That's it.

**Review by HaraldPanten — APPROVED** [review]
Manual operations can be deleted, but reconciled statement lines (with invoices or other payable/receivable accounts like customers/vendors in the starting balance) can't be deleted once they're reconciled.

IMHO, this pull request is functionally correct. LGTM.

**Comment by pedrobaeza** [comment]
OK, analyzing it, it seems congruent with the new v16 approach:

- Statement lines can be created without statement.
- You can add a new statement line in any moment, so deleting the line it's not limiting you forcing to create a new statement.

Anyway, I think we should add a protection over lines belonging to an statement, as if you remove a line, your balances won't fit and you won't be able to add the line in the same statement. What do you think? I'm thinking in statements imported from banks, where the transaction list should be unalterable at that level.

**Comment by HaraldPanten** [comment]
> OK, analyzing it, it seems congruent with the new v16 approach:
> 
> * Statement lines can be created without statement.
> * You can add a new statement line in any moment, so deleting the line it's not limiting you forcing to create a new statement.
> 
> Anyway, I think we should add a protection over lines belonging to an statement, as if you remove a line, your balances won't fit and you won't be able to add the line in the same statement. What do you think? I'm thinking in statements imported from banks, where the transaction list should be unalterable at that level.

I agree, but shouldn't we work in this new IMP/FIX in a new PR?

**Comment by lk-eska** [comment]
Can we merge this now?

**Comment by etobella** [comment]
Yes, I added a proper commit message and splitted them between both modules (Sorry for making you merge them, because they were two modules and I didn't notice that)

For the next time, remember to use proper commits message https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#L1114-L1167

As you can see, I kept your authorship and no modification on the code

If it is ok for you the applied changes, I can proceed with merge

**Comment by HaraldPanten** [comment]
But I still see 2 different commits. Is that correct, @etobella ?

**Comment by etobella** [comment]
Yes, two modules are modified.

**Comment by lk-eska** [comment]
No problem @etobella ! You can merge it.

**Review by etobella — APPROVED** [review]
/ocabot merge patch

**Comment by etobella** [comment]
Trying again

/ocabot merge patch

**Pull request merged**